File: 0934_shortestBridge/main.py
from typing import List
from collections import deque

class Solution:
    def shortestBridge(self, grid: List[List[int]]) -> int:
        N = len(grid)
        direct = [[0, 1], [0, -1], [1, 0], [-1, 0]]

        def invalid(row, col):
            return row < 0 or col < 0 or row >= N or col >= N
        
        visited = set()
        def dfs(row, col):
            if invalid(row, col) or not grid[row][col] or (row, col) in visited:
                return
            visited.add((row, col))
            for row_dir, col_dir in direct:
                dfs(row + row_dir, col + col_dir)

        def bfs():
            res = 0
            q = deque(visited)
            while q:
                length = len(q)
                for i in range(length):
                    row , col = q.popleft()
                    for row_dir, col_dir in direct:
                        curR, curC = row + row_dir, col + col_dir
                        if invalid(curR, curC) or (curR, curC) in visited:
                            continue
                        if grid[curR][curC]:
                            return res
                        q.append((curR, curC))
                        visited.add((curR, curC))
                res += 1
        
        for row in range(N):
            for col in range(N):
                if grid[row][col]:
                    dfs(row, col)
                    return bfs()
        

grid = [[0,1],[1,0]]
sol = Solution()
sol.shortestBridge(grid)

File: 0934_shortestBridge/main.cpp
#include <vector>
#include <queue>
#include <iostream>

class Solution {
private:
    std::vector<std::vector<int>> direct = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    bool inValid(std::vector<std::vector<int>>& grid, int row, int col) {
        return row < 0 || row >= grid.size() || col < 0 || col >= grid[0].size();
    }

    void dfs(std::vector<std::vector<int>>& grid, std::vector<std::vector<bool>>& visited, std::queue<std::pair<int,int>>&q, int row, int col){
        if(inValid(grid, row, col) || visited[row][col] || grid[row][col] == 0) 
            return;
        visited[row][col] = true;
        q.push({row, col});
        for(int i = 0; i < direct.size(); ++i)
            dfs(grid, visited, q, row + direct[i][0], col + direct[i][1]);
    }
public:
    int shortestBridge(std::vector<std::vector<int>>& grid) {
        std::queue<std::pair<int,int>> q;
        std::vector<std::vector<bool>> visited(grid.size(), std::vector<bool>(grid[0].size(), false));

        bool explored = false;
        for(int i = 0; i < grid.size() && !explored; ++i){
            for(int j = 0; j < grid[0].size() && !explored; ++j){
                if(grid[i][j]) {
                    dfs(grid, visited, q, i, j); 
                    explored = true;
                }
            }
        }

        int res = 0;
        while(!q.empty()){
            int curLen = q.size();
            for(int i = 0; i < curLen; ++i) {
                int row = q.front().first;
                int col = q.front().second;
                q.pop();
                for(int j = 0; j < direct.size(); ++j) {
                    int cur_row = row + direct[j][0];
                    int cur_col = col + direct[j][1];
                    if(inValid(grid, cur_row, cur_col) || visited[cur_row][cur_col])
                        continue;
                    if(grid[cur_row][cur_col] == 1)
                        return res;
                    q.push({cur_row, cur_col});
                    visited[cur_row][cur_col] = true;
                }
            }
            ++res;
        }
        return -1;
    }
};

int main() {
    std::vector<std::vector<int>> grid = {{0, 1}, {1, 0}};
    Solution sol = Solution();
    int i = sol.shortestBridge(grid);
    std::cout << i <<std::endl;
    return 0;
}